### PR TITLE
[reqstool] Revamp of the requirements tool/hook

### PR DESCRIPTION
**Pull request by truthbk**
Revamped version allows comparison of versions across repositories, local and remote. It relies on `github3.py` to access github repos. It supports:
- regular authentication
- 2FA 
- Token based authentication. 

If no authentication method is provided, it will just ignore remote repositories. Similarly, to avoid being overly invasive, if the `github3.py` library is missing, it will also ignore remote repositories. 

It supports both command-line arguments and environment variables as means to provide the arguments.

### CODE REVIEW

File: lib/tasks/ci/hooks/pre-commit.py [hunk 1]
```diff
@@ -4,62 +4,189 @@
...
+def main(args):
+    remote = local = []
+    if not len(args):
+        remote = [repo.strip() for repo in os.environ.get('REQ_REMOTES', '').split(',')]
+        local = [repo.strip() for repo in os.environ.get('REQ_LOCALS', '').split(',')]
+    elif len(args) == 1:
+        local = [repo.strip() for repo in args[0].split(',')]
+    elif len(args) == 2:
+        local = [repo.strip() for repo in args[0].split(',')]
+        remote = [repo.strip() for repo in args[1]]
```
Comment (degemer) [inline lib/tasks/ci/hooks/pre-commit.py]: Missing a `split`

File: lib/tasks/ci/hooks/pre-commit.py [hunk 2]
```diff
@@ -4,62 +4,189 @@
...
+    def get_all_requirements(self):
+        reqs = {}
+        if self.api:
+            for repo in self.remote_sources:
+                try:
+                    requirements = self.get_repo_requirements(repo)
+                    if requirements:
+                        reqs[repo] = requirements
+                except BadGithubRepo:
+                    pass
```
Comment (degemer) [inline lib/tasks/ci/hooks/pre-commit.py]: Could you add a `print` warning? It's shouldn't be completely invisible, it made things hard to debug for me

**Comment by truthbk** [comment]
@degemer forgot mentioning that, we need to install the pre-release: `pip install -pre github3.py`. Let me know if that fixes it up for you.

The idea is the same tool can be used as a pre-commit hook too. If you don't set anything it should be able to just run against the local repository. In the CI we can then customize it to run against remote repositories by setting the environment variables `rake requirements` calls this (it already works this way).

File: lib/tasks/ci/hooks/pre-commit.py (hunk 1 above)
Comment (truthbk) [inline lib/tasks/ci/hooks/pre-commit.py]: Nice catch, hadn't tested that scenario I think 🙈

File: lib/tasks/ci/hooks/pre-commit.py (hunk 2 above)
Comment (truthbk) [inline lib/tasks/ci/hooks/pre-commit.py]: Sure.

File: lib/tasks/ci/hooks/pre-commit.py [hunk 3]
```diff
@@ -4,62 +4,189 @@
...
 ERR = 1
 
-def get_files(fname):
-    matches = []
-    for root, dirnames, filenames in os.walk('.'):
-        for filename in fnmatch.filter(filenames, fname):
-            matches.append(os.path.join(root, filename))
+ENV_TOKEN = 'GHTOKEN'
+ENV_USER = 'GHUSER'
+ENV_PASS = 'GHPASS'
```
Comment (truthbk) [inline lib/tasks/ci/hooks/pre-commit.py]: No problem 👍

**Comment by degemer** [comment]
I had to first `pip uninstall github.py` and then `pip install --pre github3.py` worked. I confirm it's working, nice!
I think it's a bit too slow to be a pre-commit hook 😞 
```bash
/usr/bin/time python pre-commit.py . 'DataDog/integrations-core,DataDog/integrations-extra'
[...]
7.48 real         0.85 user         0.46 sys
```

Actually, re-reading your comment you want to run it locally on one repository to check wether or not there's a conflict inside `integrations-*`. It takes ~0.60s on my Mac which is ✅ , I think we'll have to reduce the logs though. 

Should we first focus on shipping the CI/tool part and then deal with the pre-commit part?

**Pull request merged**